digester.go

```diff
@@ -1,63 +1,227 @@
+		}
+		pipeReader.Close()
+		tarWriter.Close()
+		writeCloser.Close()
+		filterer.wg.Done()
+	}()
+	return filterer
+}
+
+// A tar digester digests an archive, modifying the headers it digests by
+// calling a specified function to potentially modify the header that it's
+// about to write.
+type tarDigester struct {
+	isOpen      bool
+	nested      digester
+	tarFilterer io.WriteCloser
+}
+
+func newTarDigester(contentType string) digester {
+	nested := newSimpleDigester(contentType)
+	digester := &tarDigester{
+		isOpen:      true,
+		nested:      nested,
+		tarFilterer: nested,
+	}
+	return digester
+}
+
+func (t *tarDigester) ContentType() string {
+	return t.nested.ContentType()
+}
+
+func (t *tarDigester) Digest() digest.Digest {
+	return t.nested.Digest()
+}
+
+func (t *tarDigester) Write(p []byte) (int, error) {
+	return t.tarFilterer.Write(p)
+}
+
+func (t *tarDigester) Close() error {
+	if t.isOpen {
+		t.isOpen = false
+		return t.tarFilterer.Close()
+	}
+	return nil
 }
 
 // CompositeDigester can compute a digest over multiple items.
 type CompositeDigester struct {
-	digesters []singleDigester
+	digesters []digester
+	closer    io.Closer
 }
 
 // Restart clears all state, so that the composite digester can start over.
 func (c *CompositeDigester) Restart() {
+	if c.closer != nil {
+		c.closer.Close()
+		c.closer = nil
+	}
 	c.digesters = nil
 }
 
-// Start starts recording the digest for a new item.  The caller should call
-// Hash() immediately after to retrieve the new io.Writer.
-func (c *CompositeDigester) Start(prefix string) {
-	prefix = strings.TrimSuffix(prefix, ":")
-	c.digesters = append(c.digesters, singleDigester{digester: digest.Canonical.Digester(), prefix: prefix})
+// Start starts recording the digest for a new item ("", "file", or "dir").
+// The caller should call Hash() immediately after to retrieve the new
+// io.WriteCloser.
+func (c *CompositeDigester) Start(contentType string) {
+	if c.closer != nil {
+		c.closer.Close()
+		c.closer = nil
+	}
+	switch contentType {
+	case "":
+		c.digesters = append(c.digesters, newSimpleDigester(""))
+	case "file", "dir":
+		digester := newTarDigester(contentType)
+		c.closer = digester
+		c.digesters = append(c.digesters, digester)
+	default:
+		panic(fmt.Sprintf(`unrecognized content type: expected "", "file", or "dir", got %q`, contentType))
+	}
 }
 
 // Hash returns the hasher for the current item.
-func (c *CompositeDigester) Hash() hash.Hash {
+func (c *CompositeDigester) Hash() io.WriteCloser {
 	num := len(c.digesters)
 	if num == 0 {
 		return nil
 	}
-	return c.digesters[num-1].digester.Hash()
+	return c.digesters[num-1]
 }
 
-// Digest returns the prefix and a composite digest over everything that's been
-// digested.
+// Digest returns the content type and a composite digest over everything
+// that's been digested.
 func (c *CompositeDigester) Digest() (string, digest.Digest) {
+	if c.closer != nil {
+		c.closer.Close()
+		c.closer = nil
+	}
```

Comment:
you do this if statement in at least 3 places now.  Worth a mini function?

Reply:
Moved that logic into a method named `closeOpenDigester()`.